---
title: "new"
output: html_document
date: "2023-05-26"
---

```{r setup, include=FALSE}
p1 <- c("tidyverse", "vegan", "BiocManager")
p2 <- c("phyloseq", "ANCOMBC", "DESeq2", "ComplexHeatmap","MicrobiotaProcess","ggsci","directlabels","ggpubr")
load_package <- function(p) {
  if (!requireNamespace(p, quietly = TRUE)) {
    ifelse(p %in% p1, 
           install.packages(p, repos = "http://cran.us.r-project.org/"), 
           BiocManager::install(p))
  }
  library(p, character.only = TRUE, quietly = TRUE)
}
invisible(lapply(c(p1,p2), load_package))


```


```{r, data}
#donor
tax_don = read.table("taxonomy_donor.txt", sep = "\t", header = T)



for(i in 2:ncol(tax_don)){
  
  tax_don[, i] = str_replace(tax_don[,i], ".*_", replacement = "")
  
}


colnames(tax_don)[1] <- "otu"
tax_don <- column_to_rownames(tax_don, "otu")

otu_don = read_tsv("otu_donor.txt") 
colnames(otu_don)[1] <- "otu"
otu_don <- column_to_rownames(otu_don, "otu")
#recipient


tax_rec = read_tsv("taxonomy_recepient.txt") %>% as.matrix()



for(i in 2:ncol(tax_rec)){
  
  tax_rec[, i] = str_replace(tax_rec[,i], ".*_", replacement = "")
  
}

colnames(tax_rec)[1] <- "otu"

rownames(tax_rec) <- tax_rec[,1]
tax_rec = tax_rec[, -1]

otu_rec = read_tsv("otu_recepient.txt")
colnames(otu_rec)[1] <- "otu"
otu_rec <- column_to_rownames(otu_rec, "otu")

#phyloseqs
tax_don = as(tax_don, "matrix")

ps_don <- phyloseq(otu_table(as.matrix(otu_don), taxa_are_rows = T), phyloseq::tax_table(tax_don))

tax_rec = as(tax_rec, "matrix")
colnames(tax_rec)[2] <- "Phylum"
ps_rec <- phyloseq(otu_table(as.matrix(otu_rec), taxa_are_rows = T), phyloseq::tax_table(tax_rec))

```

# Importing the metadata
```{r, metadata}

meta_don = read_tsv("./metadata_donor.txt") %>% select(1, 3, 5, 10:11)

meta_don = column_to_rownames(meta_don, "sample_id")

colnames(meta_don)[4] <- "vol_pool"


meta_rec = read_tsv("./metadata_recipient.txt", col_names = T ) 

meta_rec = column_to_rownames(meta_rec, "tax_id")
meta_rec = ifelse(meta_rec$Group %in% c("Control1", "Control2"), "Control", "FFT")

sample_data(ps_don) <- sample_data(meta_don) 
sample_data(ps_rec) <- sample_data(meta_rec) 


```


```{r, preprocessing, prevalance}
library(phyloseq)
#remove non-classified phyla
ps_don = subset_taxa(ps_don, !Phylum %in% c("", " ", "unclassified", "unassigned", "NA", "Unassigned" ))

# filtering based on prevalence 
prevdf <- apply(otu_table(ps_don),ifelse(taxa_are_rows(ps_don), 1, 2), function(x){sum(x>0)})
prevdf <-data.frame(ASVprev = prevdf,           #Number of the samples containing a given ASV; this shows how many samples each ASV occured.
                   TaxaAbund = taxa_sums(ps_don),  #The abundance of each ASV across all samples. 
                   phyloseq::tax_table(ps_don))
head(prevdf)
#Find out the phyla that are of mostly low-prevalence features by computing the total and average prev of features in each Phylum
junkies = plyr::ddply(prevdf, "Phylum", function(df){cbind(means = round(mean(df$ASVprev), 2), sums = round(sum(df$ASVprev),
                                         2))}) %>% mutate(sanity = ifelse(means == sums, "TRUE", "FALSE")) %>% filter(sanity == TRUE) %>% pull(Phylum)

ps_don = subset_taxa(ps_don, !Phylum %in% junkies)

```



```{r, preprocessing, abundance}
# Filtering based on abundance 

#A function to find singletones. You need to be careful about this step!
out.ASV = function(phyloseq, threshold =1, binwidth = 0.01) {
  
#Loading necessary pkgs      
  pacman::p_load(glue, tidyverse, reshape2, ggrepel, S4Vectors) # nolint
#This function requires phyloseq, tidyverse and glue packages to be loaded. 
    if (sum(colSums(otu_table(phyloseq)))/ncol(otu_table(phyloseq)) == 100 ) {#making the relative abundance table
                    rel_abund = as(t(otu_table(phyloseq)), "matrix")
    } else if (sum(colSums(otu_table(phyloseq)))/ncol(otu_table(phyloseq)) == 1) {
                    rel_abund = as(t(otu_table(phyloseq)), "matrix")
                    } else {
                    rel_abund = as(t(apply(otu_table(phyloseq), 
                    ifelse(taxa_are_rows(phyloseq), 1,2), 
                    function(x) x/sum(x))), "matrix")  
                    } 
                      
                      
                      names.single = apply(rel_abund, 1, function(x){ifelse(x == threshold, TRUE, ifelse(x == sum(x),
                      TRUE, FALSE))}) %>% reshape2::melt() %>% filter(value == TRUE) %>% dplyr::select(2) %>%
                      pull   %>% as.vector()
                      
                        
                        if (length(names.single) == 0 ) {
                        print(glue("WOW! {length(names.single)} singletones detected in this dataset"))
                        qplot.noSing = qplot(rel_abund, geom = "histogram", binwidth = binwidth, 
                        show.legend = F, main = "Frequency count of relative abundance, no singletones detected") +
                        xlab ("Relative abundance in samples") + ylab("Frequency") + theme_bw()
                            
                        
                        return(structure(list(qplot.noSing)))
                            
                        } else { 
                             
                       single.ASV = rel_abund[rownames(rel_abund) %in% names.single,]
                       single.ASV[single.ASV == 0] <- NA # A separate dataset for annotation of singletones on the barplot
                            
                       qplot.withSing = qplot(rel_abund, geom = "histogram", binwidth = binwidth, 
                       main = "Frequency count of relative abundance with singletones") +
                       geom_bar(aes(single.ASV), fill = "red",  color = NA, width = binwidth)+
                       xlab ("Relative abundance in samples") + ylab("Frequency") + 
                       geom_label_repel(aes(x = 1, y =length(rel_abund)/5), 
                       label.padding =  unit(0.55, "lines"), 
                       label = glue("{length(names.single)}\n Singletones"), color = "black") + theme_bw()
                            
                       qplot.rmSing = qplot(rel_abund[!rownames(rel_abund) %in% names.single, ], geom = "histogram",
                       binwidth = binwidth, main = "Frequency count of relative abundance without singletones") +
                       xlab ("Relative abundance in samples") + ylab("Frequency")+ theme_bw()
                            
                       print(glue('Oh no..! {length(names.single)} singletones detected in the dataset'))
                       return(structure(list(qplot.withSing, qplot.rmSing, unlist(names.single))) )
                    
                        }                        
    
                             
        }
                        
single.test = out.ASV(phyloseq = ps_rec, threshold = 1, binwidth = 0.1)
singletones = single.test[[3]] #here you can extract the names of the singletones

single.test[[1]]#to show the plot with singletones
single.test[[2]]#to show the plot without singletones



look <- list()

for (i in singletones){
  
  
  look[[i]] = ps_rec@otu_table[rownames(ps_rec@otu_table) == i,] 
  
}

#Now you can remove the singletones from your ps file as follows:
ps_rec_cl = subset_taxa(ps_rec, !taxa_names(ps_rec)%in% singletones)
ps_rec_noSingle = subset_taxa(ps_rec, !taxa_names(ps_rec)%in% singletones)

```





```{r, removing mock and negative samples}

ps_rec_cl = phyloseq::subset_samples(ps_rec_cl, !Group %in% c("N(egative)",  "P(ositive)"))

ps_rec_noSingle = phyloseq::subset_samples(ps_rec_noSingle, !Group %in% c("N(egative)",  "P(ositive)"))
```


```{r, distribution}

library(ggplot2)

count_df <- ps_rec_cl@otu_table
#non-normalized
qplot(rowSums(count_df)) + theme_bw()

#log-transformed
qplot(log(rowSums(count_df))) + theme_bw()+
  ylab("Counts per sample") +
  xlab("log-transofmed data")

#Donor
count_df_don = ps_don@otu_table
qplot(rowSums(count_df_don)) + theme_bw()

#log-transformed
qplot(log(rowSums(count_df_don))) + theme_bw()+
  ylab("Counts per sample") +
  xlab("log-transofmed data")


```

```{r, transformation}
library(DESeq2)
phyloseq_to_deseq2()
#Donor
count_df_don = as(count_df_don, "matrix")
vst_count_don = varianceStabilizingTransformation(object = count_df_don, blind = FALSE)
qplot(rowSums(vst_count_don))

otu_table(ps_don) <- otu_table(vst_count_don, taxa_are_rows  = TRUE)

#Recipient

count_df_rec = as(count_df, "matrix")
vst_count_rec = varianceStabilizingTransformation(object = count_df_rec, blind = FALSE)
qplot(rowSums(vst_count_rec))

otu_table(ps_rec_cl) <- otu_table(vst_count_rec, taxa_are_rows  = TRUE)

```


# ALpha diversity
```{r}
#Richness
##Donor
ps_don_orig <- ps_don
otu_table(ps_don_orig) <- otu_table(otu_don, taxa_are_rows = TRUE)

chao_don = estimate_richness(ps_don_orig, split = TRUE, "Chao1")


## Recipient
chao_rec = estimate_richness(ps_rec_noSingle, split = TRUE, "Chao1")


#evennes: Shannon
## Donor 
library("MicrobiotaProcess")
rar_curve_don <- ggrarecurve(obj = ps_don_orig, 
                     indexNames = c("Observe", "Shannon"),
                     chunks = 400, 
                     theme(legend.spacing.y = unit(0.02, "cm"),
                           legend.text = element_text(size = 6)), show.legend=F) +
  geom_vline(xintercept = 20000, color = "red", lty = 2, alpha = 0.6) + theme_bw()

rar_curve_don

rar_don = rarefy_even_depth(ps_don_orig, replace = FALSE, sample.size = 20000)
shan_don = estimate_richness(rar_don, split = TRUE, "shannon")

#Recipient
rar_curve_rec <- ggrarecurve(obj = ps_rec_noSingle, 
                     indexNames = c("Observe", "Shannon"),
                     chunks = 400, 
                     theme(legend.spacing.y = unit(0.02, "cm"),
                           legend.text = element_text(size = 6)), show.legend=F) +
  geom_vline(xintercept = 10000, color = "red", lty =2, alpha = 0.6) + theme_bw()

rar_curve_rec

rar_rec = rarefy_even_depth(ps_rec_noSingle, replace = FALSE, sample.size = 10000)
shan_rec = estimate_richness(rar_rec, split = TRUE, "shannon")

#BRK163 BRK174 BRK187 were removed by rearefication of 10000 du to to low reads.

alpha_don <- data.frame(Chao1 = chao_don$Chao1, Shannon = shan_don$Shannon, row.names = rownames(chao_don))

long_alpha = reshape2::melt(alpha_don) 
  
p_chao = ggplot() +
  geom_violin(data = subset(long_alpha, variable=="Chao1"), aes(x = variable, y = value), trim = F, fill = "deepskyblue") + 
  geom_boxplot(data = subset(long_alpha, variable=="Chao1"), aes(x = variable, y = value), width = 0.2) +
  geom_jitter(data = subset(long_alpha, variable=="Chao1"), aes(x = variable, y = value),  color = "black") +
  ylab("Alpha diversity") + 
  theme_bw() + facet_wrap(~ variable) +
  xlab("") + theme(axis.text.x.bottom = element_blank(), strip.text = element_text( face = "bold", size = 16, color ="white"), strip.background = element_rect(fill = "darkmagenta", color = "darkmagenta"))


p_shan = ggplot() +
  geom_violin(data = subset(long_alpha, variable=="Shannon"), aes(x = variable, y = value), trim = F, fill = "forestgreen") + 
  geom_boxplot(data = subset(long_alpha, variable=="Shannon"), aes(x = variable, y = value), width = 0.2) +
  geom_jitter(data = subset(long_alpha, variable=="Shannon"), aes(x = variable, y = value),  color = "black")+
  ylab("") + 
  theme_bw() + facet_wrap(~ variable) +
  xlab("") + theme(axis.text.x.bottom = element_blank(), strip.text = element_text( face = "bold", size = 16, color ="white"), strip.background = element_rect(fill = "darkmagenta", color = "darkmagenta"))

#library(patchwork)

p_chao + p_shan + patchwork::plot_annotation("Alpha diversity", 
                                             subtitle = "Donor gorup", caption = "Caption: I hate mosqitue bite!")



```

```{r, alpha diversity recipient}
library(ggpubr)
res_met = sample_data(rar_rec)


alpha_rec = data.frame(Chao1 = chao_rec$Chao1[rownames(chao_rec) %in% rownames(shan_rec)], Shannon = shan_rec$Shannon, row.names = rownames(chao_rec[rownames(chao_rec) %in% rownames(shan_rec),]), treat = res_met$Group ) 


long_alpha_rec = reshape2::melt(alpha_rec)
long_alpha_rec$treat <- factor(long_alpha_rec$treat, levels = c("Control", "FFT"))

long_alpha_rec %>% filter(variable == "Chao1") %>% 
  ggplot() +
  geom_violin(aes(x = treat, y = value), fill = "deepskyblue", trim = F) + 
  geom_jitter(aes(x = treat, y = value)) +
  geom_boxplot(aes(x = treat, y = value), width= 0.1) +
  facet_wrap(~ variable) + stat_compare_means(aes(x = treat, y = value), method = "t.test", label.y = 10, label.x = 0.5, color = "red") +
  theme_bw()


long_alpha_rec %>% filter(variable == "Shannon") %>% 
  ggplot() +
  geom_violin(aes(x = treat, y = value), fill = "forestgreen", trim = F) + 
  geom_jitter(aes(x = treat, y = value)) +
  geom_boxplot(aes(x = treat, y = value), width= 0.1) +
  facet_wrap(~ variable) + stat_compare_means(aes(x = treat, y = value), method = "t.test", label.y = 0.05, label.x = 0.5, color = "red") +
  theme_bw()



```

```{r, alpha: comparison}

res_cont = alpha_rec[alpha_rec$treat=="Control",]
res_fft =  alpha_rec[alpha_rec$treat=="FFT",]

meandf_chao = data.frame(don=alpha_don[rownames(alpha_don) == "BRK77", "Chao1"], rec_cont = mean(res_cont$Chao1), rec_fft = mean(res_fft$Chao1), row.names = "Chao1")

reshape2::melt(meandf_chao) %>% 
  ggplot() +
  geom_col(aes(x = variable , y = value, fill = variable)) +
  ggtitle("Comparison between donor and treatment gorups", "Chao") +
  scale_fill_manual(values = c("firebrick", "cyan4", "darkgrey")) + 
  theme_bw() + 
  stat_compare_means(aes(x = variable , y = value), method = "t.test")
  
#t.test(meandf_chao$don, meandf_chao$recon)

```



## Gloomer
#https://github.com/farhadm1990/16S-rRNA-amplicon/blob/main/R_steps.md#creating-a-stacked-barplot-for-visualizating-the-compositon-of-microbiota-in-different-taxonomic-level

```{r, taxa agglomeration Recepient }
#A function to create unique names for each ASV. It removes any NA in Order level then attempts to use the name of one level higher taxa for those 
#who have similar names, e.g. uncultured_bacterium

gloomer = function(ps = data, taxa_level = taxa_level, NArm = "TRUE"){
    rank.names = c('Kingdom','Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species')
    

#====================Sometimes in genus level, we might have multiple uncultured organisms, which if we want to make unique out of them for the species level it won't work====
    #since adding uncultured to uncultered is sill duplication. therefore if the taxa_level is set to species we first make a unique genus and then we go further to the speices===#

#Removing unculured Family
ps = subset_taxa(ps, !Family %in% c("uncultured", "NA", "uncategorized", "unassigend", "", " "))
    
if(taxa_level == "Species") {

    ps = subset_taxa(ps, !Genus %in% NA)#we remove genus tagged NA
phyloseq::tax_table(ps)[, taxa_level] <- ifelse(is.na(phyloseq::tax_table(ps)[, taxa_level]), paste0("unknown"), paste(phyloseq::tax_table(ps)[, taxa_level]))#convert NA in species into unknown
    
  physeq = tax_glom(physeq = ps, taxrank = taxa_level, NArm = NArm)
  taxdat = phyloseq::tax_table(physeq)[, seq_along(rank.names[1:which(rank.names == taxa_level)])]

   taxdat = taxdat[complete.cases(taxdat),] %>% as.data.frame
    otudat = otu_table(physeq)
   
#first take care of the uncultured genus
taxdat[,6] = ifelse(taxdat[,6] %in% c("uncategorized", NA, "uncultured", "unassigend", "", " "),
       paste0("[", taxdat[,length(rank.names[1:which(rank.names=="Genus")])-1], "]", "_", taxdat[,6]), taxdat[,6])
    
spec1 = taxdat[, taxa_level] %>% as.vector
spec2  = taxdat[, taxa_level] %>% as.vector

    uni  = matrix(NA, ncol = length(spec2), nrow = length(spec1))
    for(i in seq_along(spec1)){
        for(j in seq_along(spec2)){
    uni[i, j] = ifelse(spec1[i] == spec2[j] , "TRUE", "FALSE")
    }
        }

rownames(uni) <-spec1
colnames(uni) <- spec2   
uni[upper.tri(uni, diag = TRUE)] = 0 #get rid of diagonals and upper triangle

duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE") 

if(dim(duplis)[[1]] > 0) {
duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE") %>% dplyr::select(1) %>% unique() %>% unlist %>% as.vector
taxdat = taxdat %>% mutate( uni= ifelse(taxdat[, taxa_level] %in% duplis,
                    paste0("[", taxdat[,length(rank.names[1:which(rank.names==taxa_level)])-1], "]", "_", taxdat[,taxa_level]), taxdat[,taxa_level]))

#check if all the names are unique at species level, otherwise we will bring family instead of genus
   dupies <-  taxdat[duplicated(taxdat[,"uni"]), "uni"] 
    if(length(dupies)>0) {
        taxdat = taxdat %>% data.frame %>% mutate( uni2= ifelse(taxdat[, "uni"] %in% dupies,
                    paste0("[", taxdat[,length(rank.names[1:which(rank.names==taxa_level)])-2], "]", "_", taxdat[,"uni"]), taxdat[,"uni"]))
        
        taxdat[, taxa_level] = taxdat[, "uni2"]
        taxdat[, "uni"] <- NULL
        taxdat[, "uni2"] <- NULL
        taxdat <- as(taxdat, "matrix")   
        rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
        rownames(taxdat) <- taxdat[, taxa_level]
        taxdat <- phyloseq::tax_table(taxdat)
        taxa_names(physeq) <- taxa_names(taxdat)
        phyloseq::tax_table(physeq) <- taxdat
        otu_table(physeq) <- otudat
        
    }
    else 
    {
        
taxdat[, taxa_level] = taxdat[, "uni"]
taxdat[, "uni"] <- NULL
taxdat <- as(taxdat, "matrix")   
rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
rownames(taxdat) <- taxdat[, taxa_level]
taxdat <- phyloseq::tax_table(taxdat)
taxa_names(physeq) <- taxa_names(taxdat)
phyloseq::tax_table(physeq) <- taxdat
otu_table(physeq) <- otudat
           }
    
} else {
    
taxdat <- as.matrix(taxdat) 
taxdat <- phyloseq::tax_table(taxdat)
rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
rownames(taxdat) <- taxdat[, taxa_level]
taxdat <- phyloseq::tax_table(taxdat)
taxa_names(physeq) <- taxa_names(taxdat)
phyloseq::tax_table(physeq) <- taxdat
otu_table(physeq) <- otudat
    
}
       
    
#==========================================# 
} else if (taxa_level == "Genus") {
    
    physeq = tax_glom(physeq = ps, taxrank = taxa_level, NArm = NArm)
    taxdat = phyloseq::tax_table(physeq)[, seq_along(rank.names[1:which(rank.names == taxa_level)])]
    
   taxdat = taxdat[complete.cases(taxdat),] %>% as.data.frame
    otudat = otu_table(physeq)
    
# take care of the uncultured genus
taxdat[,6] = ifelse(taxdat[,6] %in% c("uncategorized", NA, "uncultured", "unassigend", "", " ", "XI", "group"),
       paste0("[", taxdat[,length(rank.names[1:which(rank.names==taxa_level)])-1], "]", "_", taxdat[,taxa_level]), taxdat[,taxa_level])
    
gen1 = taxdat[, taxa_level] %>% as.vector
gen2  = taxdat[, taxa_level] %>% as.vector

    uni  = matrix(NA, ncol = length(gen2), nrow = length(gen1))
    for(i in seq_along(gen1)){
        for(j in seq_along(gen2)){
    uni[i, j] = ifelse(gen1[i] == gen2[j] , "TRUE", "FALSE")
    }
        }

rownames(uni) <-gen1
colnames(uni) <- gen2   
uni[upper.tri(uni, diag = TRUE)] = 0 #get rid of diagonals and upper triangle

duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE")

        if(dim(duplis)[[1]] > 0){#if there is not duplications, we can simply use the taxa names as the row name
    
        duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE") %>% dplyr::select(1)%>% unique() %>% unlist %>% as.vector
        taxdat = taxdat %>% mutate( uni= ifelse(taxdat[, taxa_level] %in% duplis, 
                    paste0("[", taxdat[,length(rank.names[1:which(rank.names==taxa_level)])-1], "]", "_", taxdat[,taxa_level]), taxdat[,taxa_level]))
    
        taxdat[, taxa_level] = taxdat[, "uni"]
        taxdat[, "uni"] <- NULL

        taxdat <- as(taxdat, "matrix")
 
        rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
        rownames(taxdat) <- taxdat[taxdat[,taxa_level] %in% rownames(otudat), taxa_level]
        taxdat <- as.matrix(taxdat) 
        taxdat <- phyloseq::tax_table(taxdat)
        taxa_names(physeq) <- taxa_names(taxdat)
        phyloseq::tax_table(physeq) <- taxdat
        otu_table(physeq) <- otudat
 
        } else {

        taxdat <- as.matrix(taxdat) 
        taxdat <- phyloseq::tax_table(taxdat)
        rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
        rownames(taxdat) <- taxdat[, taxa_level]
        taxdat <- phyloseq::tax_table(taxdat)
        taxa_names(physeq) <- taxa_names(taxdat)
        phyloseq::tax_table(physeq) <- taxdat
        otu_table(physeq) <- otudat
       }   
    
} else {
    
    
physeq = tax_glom(physeq = ps, taxrank = taxa_level, NArm = TRUE)
    taxdat = phyloseq::tax_table(physeq)[, seq_along(rank.names[1:which(rank.names == taxa_level)])]
    
taxdat = taxdat[complete.cases(taxdat),] %>% as.data.frame
otudat = otu_table(physeq)
    
spec1 = taxdat[, taxa_level] %>% as.vector
spec2  = taxdat[, taxa_level] %>% as.vector

    uni  = matrix(NA, ncol = length(spec2), nrow = length(spec1))
    for(i in seq_along(spec1)){
        for(j in seq_along(spec2)){
    uni[i, j] = ifelse(spec1[i] == spec2[j] , "TRUE", "FALSE")
    }
        }

rownames(uni) <-spec1
colnames(uni) <- spec2   
uni[upper.tri(uni, diag = TRUE)] = 0 #get rid of diagonals and upper triangle

duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE")

if(dim(duplis)[[1]] > 0){#if there is not duplications, we can simply use the taxa names as the row name
    
    duplis = uni %>% reshape2::melt() %>% filter(value == "TRUE") %>% dplyr::select(1)%>% unique() %>% unlist %>% as.vector
taxdat = taxdat %>% mutate( uni= ifelse(taxdat[, taxa_level] %in% duplis, 
                    paste(taxdat[,length(rank.names[1:which(rank.names==taxa_level)])-1], "_", taxdat[,taxa_level]), taxdat[,taxa_level]))
    
taxdat[, taxa_level] = taxdat[, "uni"]
taxdat[, "uni"] <- NULL
taxdat <- as.matrix(taxdat)   
rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
rownames(taxdat) <- taxdat[, taxa_level]
taxdat <- phyloseq::tax_table(taxdat)
taxa_names(physeq) <- taxa_names(taxdat)
phyloseq::tax_table(physeq) <- taxdat
otu_table(physeq) <- otudat
} else {

taxdat <- as.matrix(taxdat) 
taxdat <- phyloseq::tax_table(taxdat)
rownames(otudat) <- taxdat[rownames(taxdat) %in% rownames(otudat), taxa_level]
rownames(taxdat) <- taxdat[, taxa_level]
taxdat <- phyloseq::tax_table(taxdat)
taxa_names(physeq) <- taxa_names(taxdat)
phyloseq::tax_table(physeq) <- taxdat
otu_table(physeq) <- otudat
}
#ps = phyloseq(otu_table(otudat, taxa_are_rows = T), phyloseq::tax_table(as.matrix(taxdat)), sample_data(physeq))
 

}
return(physeq) 
    }

    

```


```{r, Barplot from Phylum Recepient material}
#making barplot based from glomerated phyloseq from:
# https://github.com/farhadm1990/16S-rRNA-amplicon/blob/main/R_steps.md


phylcol=c('darkgreen', "gold",'darkmagenta','cadetblue2', 'cornflowerblue', 'plum4',
          'darkgoldenrod3','aquamarine4', 'cadetblue2', 'darkgreen', 'mediumseagreen', 'red','Gray',
        'steelblue2','darkmagenta')

phylcol

#Agglomeration
phyl_rec <- gloomer(rar_rec, "Phylum", NArm = TRUE) 

 



#Merging counts for treatments
merged_phyl_rec <- merge_samples(phyl_rec, "Group") 

merged_phyl_rec = transform_sample_counts(merged_phyl_rec, function(x){x/sum(x)*100}) 



sample_data(merged_phyl_rec)$Group <- factor(rownames(sample_data(merged_phyl_rec)), levels = c("Control","FFT"))
                                     
#Barplot                                    
p_rec = plot_bar(merged_phyl_rec, fill="Phylum", x = "Group") + 
  geom_col(color = "white", show.legend = F) +
  scale_fill_manual(values = list("Actinobacteriota" = "aquamarine4",
     "Bacteroidota" = "gold",
     "Firmicutes" = "red",
     "Proteobacteria" = "darkgoldenrod3")) + 
    xlab("Treatments") + ylab("") + 
scale_y_continuous(labels = function(x) format(x, scientific = F), n.breaks = 10)+
  #to print the scale in scientific notation
            theme_bw() + 
            theme(text = element_text(size =15, face = "bold"))

ggsave(plot = p_rec, "./outputs/rec_stack_phylum.jpeg", device = "jpeg", height = 8, width = 7)

```

```{r, taxa agglomeration Phylom Donor}
phylcol=c('darkgreen', "gold",'darkmagenta','cadetblue2', 'cornflowerblue', 'plum4',
          'darkgoldenrod3','aquamarine4', 'cadetblue2', 'darkgreen', 'mediumseagreen', 'red','Gray',
        'steelblue2','darkmagenta')
list("Actinobacteriota" = "aquamarine4",
     "Bacteroidota" = "gold",
     "Firmicutes" = "red",
     "Proteobacteria" = "darkgoldenrod3")
#Agglomeration
phyl_don <- gloomer(rar_don, "Phylum", NArm = TRUE) 
pooled_phyl_don = subset_samples(phyl_don, Extract_name == "Pooled")
phyl_don@tax_table
pooled_phyl_don = transform_sample_counts(pooled_phyl_don, fun = function(x){x/sum(x)*100}) 

p_don = plot_bar(pooled_phyl_don, fill="Phylum") + 
  geom_col(color = "white") +
  scale_fill_manual(values = list("Actinobacteriota" = "aquamarine4",
     "Bacteroidota" = "gold",
     "Firmicutes" = "red",
     "Proteobacteria" = "darkgoldenrod3")) + 
    xlab("Pooled donor") + ylab("Relative abundance, %") + 
scale_y_continuous(labels = function(x) format(x, scientific = F), n.breaks = 10)+
  #to print the scale in scientific notation
            theme_bw() + 
            theme(text = element_text(size =15, face = "bold")) + 
  guides(fill = guide_none())

ggsave(plot = p_don, "./outputs/don_stack_phylum.jpeg", device = "jpeg", height = 8, width = 6)

# Patchwork

p_don + p_rec + plot_layout(widths = 1) #+ plot_annotation("I am a title", "I am a legend", "I am a caption")

ggsave("./outputs/phyl_stack_compare.jpeg", device = "jpeg", width = 8, height = 10, dpi = 300)

colors()
RColorBrewer::brewer.pal.info
```

```{r, Family level pooled donor and recipient barplot}
fam_colors <- c("#FF0000", "#00FF00", "#0000FF", "#FFFF00", "#FF00FF", "#00FFFF", "#800000", "#008000", 
            "#000080", "#808000", "#800080", "#008080", "#808080", "#FFA500", "#800080", "#FFC0CB", 
            "#FF69B4", "#4B0082", "#ADD8E6", "#FF7F50", "#FFD700", "#9ACD32", "#7FFFD4", "#DC143C", 
            "#00BFFF", "#32CD32", "#FF8C00", "#9932CC", "#00FA9A", "#FF4500")

print(colors)

#Agglomeration
fam_rec <- gloomer(rar_rec, "Family", NArm = TRUE) 

fam_don <- gloomer(rar_don, "Family", NArm = TRUE) 
pooled_fam_don = subset_samples(fam_don, Extract_name == "Pooled")
pooled_fam_don@tax_table
pooled_fam_don = transform_sample_counts(pooled_fam_don, fun = function(x){x/sum(x)*100})

#taking care of wierd taxa names
temp_taxa = pooled_fam_don@tax_table %>% data.frame() %>% mutate(new_fam = ifelse(Family %in% c("XI" , "group"),  glue("[{Order}]_{Family}"),  Family))  %>% mutate(Family =NULL, Family = new_fam, new_fam = NULL) 

rownames(temp_taxa) <- temp_taxa$Family
temp_taxa = as(temp_taxa, 'matrix')


pooled_fam_don@tax_table <- phyloseq::tax_table(temp_taxa)

#Merging counts for treatments
merged_fam_rec <- merge_samples(fam_rec, "Group") 

merged_fam_rec = transform_sample_counts(merged_fam_rec, function(x){x/sum(x)*100}) 
merged_fam_rec@sam_data$Group <- rownames(merged_fam_rec@sam_data)

#creating barplot from recipient
f_rec = plot_bar(merged_fam_rec, fill="Family", x = "Group") + 
  geom_col(color = NA, show.legend = F) +
  scale_fill_manual(values = fam_colors) + 
    xlab("Treatments") + ylab("") + 
scale_y_continuous(labels = function(x) format(x, scientific = F), n.breaks = 10)+
  #to print the scale in scientific notation
            theme_bw() + 
            theme(text = element_text(size =15, face = "bold"))

#creating barplot from donor material
f_rec = plot_bar(merged_fam_rec, fill="Family", x = "Group") + 
  geom_col(color = NA, show.legend = F) +
  scale_fill_manual(values = fam_colors) + 
    xlab("Treatments") + ylab("") + 
scale_y_continuous(labels = function(x) format(x, scientific = F), n.breaks = 10)+
  #to print the scale in scientific notation
            theme_bw() + 
            theme(text = element_text(size =15, face = "bold"))



```

```{r}

ps_don_phyl = gloomer(ps = rar_don, taxa_level = "Phylum", NArm = TRUE)
ps_don_gen = gloomer(ps = rar_don, taxa_level = "Genus", NArm = TRUE)

ps_rec_phyl = 
```



```{r}


col_fil <- pal_jco("default")(10)

col_scale <- scale_color_jco()
```

#Build phyloseq object
```{r pressure, echo=FALSE}
otu_donor <- read.table(file = "otu_PrePhage3_donor_20230525.txt", sep = "\t", header = T, row.names = 1, 
                  skip = 0, comment.char = "")
taxonomy_donor <- otu_donor[,"Taxonomy", drop = FALSE]
otu_donor <- subset(otu_donor, select = -Taxonomy)

# clean the taxonomy, Greengenes format
tax_donor <- taxonomy_donor %>%
  select(Taxonomy) %>% 
  separate(Taxonomy, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), ";")


tax_donor.clean <- data.frame(row.names = row.names(tax_donor),
                        Kingdom = str_replace(tax_donor[,1], "k__",""),
                        Phylum = str_replace(tax_donor[,2], "p__",""),
                        Class = str_replace(tax_donor[,3], "c__",""),
                        Order = str_replace(tax_donor[,4], "o__",""),
                        Family = str_replace(tax_donor[,5], "f__",""),
                        Genus = str_replace(tax_donor[,6], "g__",""),
                        Species = str_replace(tax_donor[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax_donor.clean[is.na(tax_donor.clean)] <- ""
tax_donor.clean[tax_donor.clean=="__"] <- ""

for (i in 1:nrow(tax_donor.clean)){
  if (tax_donor.clean[i,7] != ""){
    tax_donor.clean$Species[i] <- paste(tax_donor.clean$Genus[i], tax_donor.clean$Species[i], sep = " ")
  } else if (tax_donor.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax_donor.clean[i,1], sep = " ")
    tax_donor.clean[i, 2:7] <- kingdom
  } else if (tax_donor.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax_donor.clean[i,2], sep = " ")
    tax_donor.clean[i, 3:7] <- phylum
  } else if (tax_donor.clean[i,4] == ""){
    class <- paste("Unclassified", tax_donor.clean[i,3], sep = " ")
    tax_donor.clean[i, 4:7] <- class
  } else if (tax_donor.clean[i,5] == ""){
    order <- paste("Unclassified", tax_donor.clean[i,4], sep = " ")
    tax_donor.clean[i, 5:7] <- order
  } else if (tax_donor.clean[i,6] == ""){
    family <- paste("Unclassified", tax_donor.clean[i,5], sep = " ")
    tax_donor.clean[i, 6:7] <- family
  } else if (tax_donor.clean[i,7] == ""){
    tax_donor.clean$Species[i] <- paste("Unclassified ",tax_donor.clean$Genus[i], sep = " ")
  }
}

metadata_donor <- read.table("PrePhage3_donor_20230525.txt", sep="\t", header=TRUE, row.names = 1)


OTU_d = otu_table(as.matrix(otu_donor), taxa_are_rows = TRUE)
TAX_d = phyloseq::tax_table(as.matrix(tax_donor.clean))
SAMPLE_d <- sample_data(metadata_donor)
# TREE = read_tree("tree.nwk")
# merge the data
ps_donor <- phyloseq(OTU_d, TAX_d, SAMPLE_d)

saveRDS(ps_donor, "ps.test.dbrda.rds")
```
#tax
```{r}
tax_donor
```

#tax clean
```{r}
tax_donor.clean
```

# Rarefaction curve
```{r}
# Extract the OTU table from the phyloseq object
mat_donor <- t(otu_table(ps_donor))

class(mat_donor) <- "matrix"
class(mat_donor)
# Calculate the total number of reads (samples)
raremax_donor <- min(rowSums(mat_donor))

# Create the rarefaction curve
rarecurve(mat_donor, step = 100, sample = raremax_donor, col = "blue", label = TRUE)
```


```{r}
set.seed(111) # keep result reproductive
ps.rarefied_donor = rarefy_even_depth(ps_donor, rngseed=1, sample.size=1000, replace=F)
meta_donor_test <- sample_data(ps.rarefied_donor)
meta_donor_test$Extract_name <- sub("FFT2","FFT",meta_donor_test$Extract_name)
meta_donor_test$Extract_name <- sub("FFT1","FFT",meta_donor_test$Extract_name)
meta_donor_test$Extract_name <- sub("Control2","Control",meta_donor_test$Extract_name)
meta_donor_test$Extract_name <- sub("Control1","Control",meta_donor_test$Extract_name)
sample_data(ps.rarefied_donor)<-meta_donor_test
saveRDS(ps.rarefied_donor, "ps.rarefied_donor.test.dbrda.rds")
#meta_test <- sample_donor_data(ps.rarefied_donor)
#meta_test$Group <- sub("FFT2","FFT",meta_test$Group)
#meta_test$Group <- sub("FFT1","FFT",meta_test$Group)
#meta_test$Group <- sub("Control2","Control",meta_test$Group)
#meta_test$Group <- sub("Control1","Control",meta_test$Group)
#sample_data(ps.rarefied)<-meta_test
#print(ps.rarefied)
#print(meta_test)
colSums(otu_table(ps.rarefied_donor))
ggplot2::qplot(rowSums(otu_table(ps.rarefied_donor)))
ggplot2::qplot(log(rowSums(otu_table(ps_donor))), binwidth = 0.1)
DESeq2::varianceStabilizingTransformation(otu_table(ps_donor))
```


```{r}

bac_donor.phyl <- tax_glom(ps_donor, "Genus", NArm = FALSE)

ps0_donor <- transform_sample_counts(bac_donor.phyl, function(x) x / sum(x))

#Create melted dataframe
df_donor <- psmelt(ps0_donor)

#Select last non-empty taxonomic rank
df_donor[df_donor==""] <- NA

df_donor$tax_donor <- apply(df_donor, 1, function(x) tail(na.omit(x), 1))

#Arrange samples by mean abundance
top_donor <- df_donor %>%
  group_by(Sample, Extract_name, tax_donor) %>%
  dplyr::summarize(Mean = mean(Abundance)) %>%
  arrange(-Mean)
#Show top
#top

top10_donor <- top_donor$tax_donor[1:150]
df0_donor <- df_donor %>%
  mutate(tax_donor = fct_other(tax_donor, c(as.matrix(top10_donor))))

df0_donor <- df0_donor[order(df0_donor$Sample, decreasing = TRUE), ]


barplot.all.samples <- ggplot(df0_donor, aes(Sample, Abundance, fill = fct_reorder(tax_donor, Abundance, .fun = function(x) mean(x)))) +
  geom_col(width = 0.8) +
  facet_wrap(~ Extract_name) +
  scale_fill_manual(values = rep(col_fil, 5), name = "Genus") +
  theme_classic() +
  theme(
    text = element_text(size = 10, colour = "Black"),
    axis.line = element_line(size = 0.5),
    axis.text = element_text(size = 10, colour = "Black"),
    axis.ticks = element_line(size = 1, colour = "Black"),
    strip.background = element_rect(colour = "white", fill = "white"),
    axis.text.x = element_text(angle = 45, hjust = 1),
    strip.text.x = element_text(angle = 0, size = 12, face = "bold")
  ) +
  ylab("Relative abundance")

print(barplot.all.samples)


barplot.all.samples

```



```{r}

bac_donor.phyl <- tax_glom(ps_donor, "Genus", NArm = FALSE)

ps0_donor <- transform_sample_counts(bac_donor.phyl, function(x) x / sum(x))

#Create melted dataframe
df_donor <- psmelt(ps0_donor)

#Select last non-empty taxonomic rank
df_donor[df_donor==""] <- NA

df_donor$tax_donor <- apply(df_donor, 1, function(x) tail(na.omit(x), 1))

#Arrange samples by mean abundance
top_donor <- df_donor %>%
  group_by(Sample, Extract_name,tax_donor) %>%
  dplyr::summarize(Mean = mean(Abundance)) %>%
  arrange(-Mean)
#Show top
#top

top10_donor <- top_donor$tax_donor[1:150]
df0_donor <- df_donor %>%
  mutate(tax_donor = fct_other(tax_donor, c(as.matrix(top10_donor))))

df0_donor <- df0_donor[order(df0_donor$Sample, decreasing = TRUE), ]

barplot.all.samples <- ggplot(df0_donor, aes(Sample, Abundance, fill = fct_reorder(tax_donor, Abundance, .fun = mean))) + 
  geom_col(width = 0.8) +
  # facet_wrap("Litter")+
  scale_fill_jco(name = "Taxonomy") +
  scale_fill_manual(values=rep(col_fil,5),name="Genus")+
  theme_classic() +
  theme(text = element_text(size = 10, colour = "Black"),
        axis.line=element_line(size=0.5),
        #panel.border = element_blank(),
        axis.text=element_text(size = 10, colour = "Black"),
        axis.ticks=element_line(size=1, colour = "Black"),
        strip.background = element_rect(colour = "white", fill = "white"),
        axis.text.x=element_text(angle = 45, hjust = 1),
        strip.text.x = element_text(angle = 0, size=12, face = "bold"),
        #legend.position = "none"
  ) +
  ylab("Relative abundance")


barplot.all.samples

```

```{r}
library(ggplot2)
library(forcats)
library(dplyr)

# Plot 1 - df0_donor
bac_donor.phyl <- tax_glom(ps_donor, "Genus", NArm = FALSE)
ps0_donor <- transform_sample_counts(bac_donor.phyl, function(x) x / sum(x))
df_donor <- psmelt(ps0_donor)
df_donor[df_donor==""] <- NA
df_donor$tax_donor <- apply(df_donor, 1, function(x) tail(na.omit(x), 1))
top_donor <- df_donor %>%
  group_by(Sample, Extract_name, tax_donor) %>%
  dplyr::summarize(Mean = mean(Abundance)) %>%
  arrange(-Mean)
top10_donor <- top_donor$tax_donor[1:150]
df0_donor <- df_donor %>%
  mutate(tax_donor = fct_other(tax_donor, c(as.matrix(top10_donor))))
df0_donor <- df0_donor[order(df0_donor$Sample, decreasing = TRUE), ]

# Plot 2 - df0
bac.phyl <- tax_glom(ps, "Genus", NArm = FALSE)
ps0 <- transform_sample_counts(bac.phyl, function(x) x / sum(x))
df <- psmelt(ps0)
df[df==""] <- NA
df$tax <- apply(df, 1, function(x) tail(na.omit(x), 1))
top <- df %>%
  group_by(Reference_ID, Litter, tax) %>%
  dplyr::summarize(Mean = mean(Abundance)) %>%
  arrange(-Mean)
top10 <- top$tax[1:150]
df0 <- df %>%
  mutate(tax = fct_other(tax, c(as.matrix(top10))))
df0 <- df0[order(df0$Sample, decreasing = TRUE), ]

# Combine plots
combined_plot <- ggplot() +
  # Plot 1 - df0_donor
  geom_col(data = df0_donor, aes(Sample, Abundance, fill = fct_reorder(tax_donor, Abundance, .fun = mean)), width = 0.8) +
  # Plot 2 - df0
  geom_col(data = df0, aes(Reference_ID, Abundance, fill = fct_reorder(tax, Abundance, .fun = mean)), width = 0.8) +
  # Additional styling
  scale_fill_manual(values = rep(col_fil, max(length(unique(df0_donor$tax_donor)), length(unique(df0$tax)))), name = "Genus") +
  theme_classic() +
  theme(
    text = element_text(size = 10, colour = "Black"),
    axis.line = element_line(size = 0.5),
    axis.text = element_text(size = 10, colour = "Black"),
    axis.ticks = element_line(size = 1, colour = "Black"),
    strip.background = element_rect(colour = "white", fill = "white"),
    axis.text.x = element_text(angle = 45, hjust = 1),
    strip.text.x = element_text(angle = 0, size = 12, face = "bold"),
    legend.position = "none"
  ) 

combined_plot


```

